### PR TITLE
Signup: Add signup flow for upgrading a domain-only site

### PR DESCRIPTION
**Pull request by yurynix**
# This is a replacement for #11224 which was broken for the main flow.
It is different only by the last commit.

This PR aims to allow users to upgrade their "domain only site" to a regular site with plan.
This PR also fixes a few bugs related to previous "featuregate" work ( the first 2 commits ).

#### Testing instructions
  
Run `git checkout add/upgrade-domain-only` and start your server, or open a [live branch](https://calypso.live/?branch=add/upgrade-domain-only)

Flow:
0. Click "Create a new site" on a "domain only site" ( you might want to create one of those first ).
screen shot 2017-02-07 at 14 30 17

1. Choose a theme
screen shot 2017-02-07 at 14 28 01

2. Choose a plan ( no free plan )
screen shot 2017-02-07 at 14 25 26

3. Checkout ( with discounted price due to existing domain )
screen shot 2017-02-07 at 14 01 11

4. Thank you screen AND Site has a plan + no `is_domain_only` option, that requires the server side patch ( D4297 ).
screen shot 2017-02-07 at 16 48 03

5. Start over again and make sure sign up flow work for default use case and free .wordpress.com subdomain.

#### Reviews
  
- [ ] Code
- [ ] Product

### CODE REVIEW

**Comment by gziolo** [comment]
Even though code changes look good, there are some issues with the flow:

domain-only-upgrade

There is also wrong page displayed instead of Thank You page:
screen shot 2017-02-13 at 11 44 57

**Comment by drewblaisdell** [comment]
Disabling the cart during signup fixes that issue. Cart synchronization isn't something we need to worry about during signup, and what little interaction between signup and the cart there is is managed by `SignupCart`. If you're cool with my changes, the code LGTM.

File: client/my-sites/controller.js [hunk 1]
```diff
@@ -145,26 +145,26 @@ function isPathAllowedForDomainOnlySite( pathname, domainName ) {
 
 function onSelectedSiteAvailable( context ) {
 	const selectedSite = sites.getSelectedSite();
-	const state = context.store.getState();
+	const getState = () => context.store.getState();
```
Comment (Tug) [inline client/my-sites/controller.js]: This shouldn't be needed, you get a reference to the store here so even if `setSelectedSiteId` modifies it you'll be able to read the updated value

Comment (yurynix) [inline client/my-sites/controller.js]: `getState()` returns a different object after a state transition after `dispatch` that happens later in this method.

If you meant that you can call each time `context.store.getState()`, that is right, but it's long, i prefer just write `getState()` if you don't mind =)

Comment (Tug) [inline client/my-sites/controller.js]: Oh you're right nevermind then

**Comment by drewblaisdell** [comment]
@yurynix and I discussed this on Slack and decided to move the `CartStore`/`CartSynchronizer` refactor to a separate PR to make both this and that easier to review.

**Comment by yurynix** [comment]
I'm cool with moving that forward for now 👍  Thank you all for your input, thank you @drewblaisdell for helping solving this one. 

After @drewblaisdell 's commit this works fine for me.

Note: there is a race condition in checkout ( between site fetch and empty cart redirect ) that is solved in another PR, if you sometimes get no sites after checkout that's ok ( for this PR ).

**Comment by drewblaisdell** [comment]
I tested:

- Creating a free site in the `main` flow.
- Creating a free site with a plan in the `main` flow.
- Buying a domain/plan in the `main` flow.
- The testing instructions in the PR description.

:+1:

**Pull request merged**